Add Correction class

gonzaponte opened the pull request:
An interface for accessing geometrical corrections.

gonzaponte committed:
Add Correction class

An interface for accessing geometrical corrections.

jjgomezcadenas reviewed (commented):
The code is elegant and concise but difficult to read (at least for this reviewer). Could you please add comments?

gonzaponte committed:
Comments

gonzaponte committed:
tests

gonzaponte committed:
switch to relative imports

gonzaponte committed:
test with toy data

gonzaponte committed:
use namedtuples

gonzaponte committed:
improve tests

gonzaponte committed:
better coding for Correction

gonzaponte committed:
rename variable

gonzaponte committed:
default values added

gonzaponte committed:
refactoring

gonzaponte committed:
update tests

gonzaponte committed:
interpolation strategy added

gonzaponte committed:
move to a new file

gonzaponte committed:
add Measurement namedtuple

gonzaponte committed:
internal method added

gonzaponte committed:
Add fcorrection

gonzaponte committed:
change flatten policy

gonzaponte committed:
fix Fcorrection

gonzaponte committed:
tests made with hypothesis

gonzaponte committed:
recover old tests + names consistency

jjgomezcadenas committed:
Mark private attributes with underscore

The interface of this class consist of the constructor and the
__call__ method. While the rest of the methods in the class were
marked with _ to indicate private-to-the-class, the attributes were
not.

The test suite contains tests which go around the public
interface (e.g, they test private attributes that start with _). This
may be not needed if correct tests of the public interface are
implemented. Furthermore, tests act as documentation. By using private
attributes in the test one encourages clients to ignore the interface.

We need to check whether these tests are necessary.

jjgomezcadenas committed:
Remove superflous parameter u_pars

Fcorrection define functions f and u_f both of them depending on the
same parameter(s) specfied by pars. There is no user case in the tests
showing that an additional parameter(s) u_pars is needed. Thus, we
have removed the u_pars as superfluous.

If there is a genuine need for this parameter, demonstrate it with a
test.

jjgomezcadenas committed:
Move correction instantiations out of hypothesis generators

jjgomezcadenas committed:
Add flaky to chi2 test

jjgomezcadenas committed:
Move __call__ after __init__

__call__ is part of the interface and must be highlighted rather than
being in the bottom behind the private methods.

jjgomezcadenas committed:
Minor fixes

gonzaponte committed:
pass

jjgomezcadenas reviewed (dismissed):
Class Correction accepts a matrix of energies (or previously existing factors) used to produce normalisation factors for the energy. The class comes with extensive testing and has been thoroughly revised by @jacg  and @jjgomezcadenas

gonzaponte commented on invisible_cities/reco/corrections_test.py:
```diff
@@ -154,35 +155,40 @@ def test_correction_attributes_2d(toy_data_2d):
-    assert_allclose(F_test, F )
-    assert_allclose(U_test, Fu)
+    # create a collection of (x,y) point such that the
+    # x coordinates are stored in X_sample and the y coordinates in Y_sample
```
Just so I can keep my training as a padawan... why is this better than using np.repeat and np.tile?
